src/engine/routing_algorithms/many_to_many.cpp

```diff
@@ -120,27 +121,65 @@ addLoopWeight(const DataFacade<mld::Algorithm> &, const NodeID, EdgeWeight &, Ed
     return false;
 }
 
-template <bool DIRECTION>
+template <typename MultiLevelPartition>
+inline LevelID getNodeQueryLevel(const MultiLevelPartition &partition,
+                                 NodeID node,
+                                 const PhantomNode &phantom_node)
+{
+    auto highest_diffrent_level = [&partition, node](const SegmentID &phantom_node) {
+        if (phantom_node.enabled)
+            return partition.GetHighestDifferentLevel(phantom_node.id, node);
+        return INVALID_LEVEL_ID;
+    };
+
+    return std::min(highest_diffrent_level(phantom_node.forward_segment_id),
+                    highest_diffrent_level(phantom_node.reverse_segment_id));
+}
+
+template <typename MultiLevelPartition>
+inline LevelID getNodeQueryLevel(const MultiLevelPartition &partition,
+                                 NodeID node,
+                                 const std::vector<PhantomNode> &phantom_nodes,
+                                 const std::size_t phantom_index,
+                                 const std::vector<std::size_t> &phantom_indices)
+{
+    auto level = [&partition, node](const SegmentID &source, const SegmentID &target) {
+        if (source.enabled && target.enabled)
+            return partition.GetQueryLevel(source.id, target.id, node);
+        return INVALID_LEVEL_ID;
+    };
+
+    const auto &source_phantom = phantom_nodes[phantom_index];
+
+    auto result = INVALID_LEVEL_ID;
+    for (const auto &index : phantom_indices)
+    {
+        const auto &target_phantom = phantom_nodes[index];
+        auto min_level = std::min(
+            std::min(level(source_phantom.forward_segment_id, target_phantom.forward_segment_id),
```

Comment:
It seems to me that this could be implemented a bit more performant, since the level between `node` and `source_phantom` does not change over the loop. Depending on how often this is called, it might be a small performance bottle-neck.

Also, we could filter this list for vertices within the same cell to eliminate duplicate checks.

Reply:
ok, i will update here